### PR TITLE
plugins/neocord: init

### PR DESCRIPTION
**Pull request by redyf**
Add support for https://github.com/IogaMaster/neocord

### CODE REVIEW

**Comment by traxys** [comment]
You are going to need to specify to not run the tests, as they try to acess the network. Search for dontRun in the tests for exemples

**Comment by redyf** [comment]
> You are going to need to specify to not run the tests, as they try to acess the network. Search for dontRun in the tests for exemples

Thanks! I'll check

**Review by GaetanLepage — CHANGES_REQUESTED** [review]
Thank you very much for this great looking implementation.

However, I am sorry but will ask you to use our new standard for packaging neovim plugins.
You can find an example [here](https://github.com/nix-community/nixvim/blob/main/plugins/lsp/nvim-lightbulb.nix).
It should be pretty straightforward to do the conversion though.

**Comment by redyf** [comment]
> Thank you very much for this great looking implementation.
> 
> However, I am sorry but will ask you to use our new standard for packaging neovim plugins. You can find an example [here](https://github.com/nix-community/nixvim/blob/main/plugins/lsp/nvim-lightbulb.nix). It should be pretty straightforward to do the conversion though.

No worries, I'll implement it.

File: plugins/utils/neocord.nix [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        "editingTimer"
+        "fileExplorerText"
+        "gitCommitText"
+        "pluginManagerText"
+        "readingText"
+        "workspaceText"
+        "lineNumberText"
+        "terminalText"
+      ];
+
```
Comment (GaetanLepage) [inline plugins/utils/neocord.nix:46]: You can remove all that. This creates deprecation errors for those options.
But, as this module has never existed in nixvim, there is no need for warning the user about deprecated options.

File: plugins/utils/neocord.nix [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      debounce_timeout = helpers.defaultNullOpts.mkInt 10 ''
+        Number of seconds to debounce events.
+        (or calls to `:lua package.loaded.neocord:update(<filename>, true)`)
+      '';
+
+      enable_line_number = helpers.defaultNullOpts.mkBool false ''
+        Displays the current line number instead of the current project.
+      '';
+
+      blacklist = helpers.defaultNullOpts.mkNullable (types.listOf types.str) "[]" ''
```
Comment (GaetanLepage) [inline plugins/utils/neocord.nix:52]:
```suggestion
      blacklist = helpers.defaultNullOpts.mkListOf types.str "[]" ''
```

File: plugins/utils/neocord.nix [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        helpers.defaultNullOpts.mkNullable
+        (types.either helpers.nixvimTypes.rawLua
+          (types.listOf (types.submodule {
+            options = {
+              label = helpers.mkNullOrOption types.str "";
+              url = helpers.mkNullOrOption types.str "";
+            };
+          })))
+        "[]"
+        ''
```
Comment (GaetanLepage) [inline plugins/utils/neocord.nix:67]:
```suggestion
        (with types;
          either
          helpers.nixvimTypes.rawLua
          (listOf (submodule {
            options = {
              label = helpers.mkNullOrOption str "";
              url = helpers.mkNullOrOption str "";
            };
          })))
        "[]"
        ''
```

File: plugins/utils/neocord.nix [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+          };
+        ```
+      '';
+
+      show_time = helpers.defaultNullOpts.mkBool true "Show the timer.";
+
+      global_timer = helpers.defaultNullOpts.mkBool false "if set true, timer won't update when any event are triggered.";
+
+      # Rich presence text options.
+      editing_text = helpers.defaultNullOpts.mkNullable (types.either types.str helpers.nixvimTypes.rawLua) "Editing %s" ''
```
Comment (GaetanLepage) [inline plugins/utils/neocord.nix:98]:
```suggestion
      editing_text = helpers.defaultNullOpts.mkStr "Editing %s" ''
```

`rawType` is accepted by default: you have nothing to do :)
See [here](https://github.com/nix-community/nixvim/blob/22351f4628ddd1556c2a35b5ca9b70405f1626ac/lib/options.nix#L138) (`maybeRaw`).

Comment (redyf) [reply plugins/utils/neocord.nix:98]: I had no idea, thanks for sharing :+1:

File: plugins/utils/neocord.nix [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        String rendered when `enableLineNumber` is set to `true` to display the current line number.
+        Can also be a lua function:
+        `function(line_number: number, line_count: number): string`
+      '';
+
+      terminal_text = helpers.defaultNullOpts.mkNullable (types.either types.str helpers.nixvimTypes.rawLua) "Using Terminal" ''
+        Format string rendered when in terminal mode.
+      '';
+    };
+    settingsExample = {
```
Comment (GaetanLepage) [inline plugins/utils/neocord.nix:144]:
```suggestion
    };

    settingsExample = {
```

**Comment by redyf** [comment]
@GaetanLepage Applied all changes, thank you so much for the help :smile:

File: tests/test-sources/plugins/utils/neocord.nix [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        git_commit_text = "Committing changes";
+        plugin_manager_text = "Managing plugins";
+        reading_text = "Reading %s";
+        workspace_text = "Working on %s";
+        line_number_text = "Line %s out of %s";
+        terminal_text = "Using Terminal";
+      };
+    };
+  };
+  example = {
```
Comment (GaetanLepage) [inline tests/test-sources/plugins/utils/neocord.nix:39]:
```suggestion
  };

  example = {
```

Last one for sure !

**Review by GaetanLepage — APPROVED** [review]
LGTM !
Thanks for having applied all the corrections !

**Pull request merged**